File: Tree/100.cpp
#include <iostream>
using namespace std;


 //Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };
 
//Liu 
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
    	if(p == NULL && q == NULL) return true;
        if(p && q) {
           if ( p -> val == q -> val 
           	    && isSameTree(p -> left, q -> left) 
           	    && isSameTree(p -> right, q -> right)){
           		//cout << "haha" << endl;
           		return true;
           }
              //return true;
           else return false;
       } 
       else return false; // if(p || q){return false;}
        
    }
};

//Find online, dfs
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
      if(!p && !q) return true;  // if(p == NULL && q == NULL){return true;}
      if(!p || !q) return false;  
      return (p->val == q->val) &&  
           isSameTree(p->left, q->left) &&   
           isSameTree(p->right, q->right);      
        
    }
};


class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
   	  if(p == NULL && q == NULL) return true;  
      if(p && q) { 
      return (p->val == q->val) &&  
           isSameTree(p->left, q->left) &&   
           isSameTree(p->right, q->right); 
      }
      if(p || q){return false;}
    }
};

//my code, 4/26/2016
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p == NULL && q == NULL) {
            return true;
        }
        if(p && q) {
            if(p -> val == q -> val && isSameTree(p -> left, q -> left) && isSameTree(p -> right, q -> right)) {
                return true;
            }  else {
                return false;
            }
        }
        return false;
    }
};


int main(){
	/*int aa = 2;
	int bb = 0;
	if(bb && (aa = 3)){

	}
	cout << aa << endl;*/

	TreeNode *a = new TreeNode(0);
	TreeNode *b = new TreeNode(0);
	a->left = new TreeNode(-5);
	b->left = new TreeNode(-8);
	Solution sol;
	cout << sol.isSameTree(a,b) << endl;

	return 0;
}

File: Tree/100.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

#my code, 10/11/2016
class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if (not p and not q):
            return True
        if (p and q):
            if (p.val == q.val and self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)):
                return True
            else:
                return False
        return False;


#Find online, Indentation error ????
class Solution(object):
    def isSameTree(self, p, q):   
	    if p and q:
	    	return p.val == q.val and self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
	    return p is q


#Find online, correct
class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        return p and q and p.val == q.val and all(map(self.isSameTree, (p.left, p.right), (q.left, q.right))) or p is q